design/k8s_secrets_labels_design.md

```diff
@@ -0,0 +1,218 @@
+
+## Issue description
+
+At current, the Secrets Provider code looks at the `K8S_SECRETS` environment variable in each of the pod manifests to know which K8s Secret to update with Conjur values. 
+
+Listing each K8s Secret in this way is not a scalable solution, especially for customers with hundreds of K8s Secrets. 
+
+## Solution
+
+We will offer the customer an *additional* option to define which K8s Secrets values are hosted in Conjur by defining them in terms of a label instead of individually. They will define the label that is attached to the K8s Secret and add it under `K8S_SECRETS_LABEL` environment variable in the Secrets Provider manifest. 
+
+The Secrets Provider will then make a separate API call to [K8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core)/[OC](https://docs.openshift.com/container-platform/4.4/rest_api/index.html#list-14), fetching all K8s Secrets with that label. We will use that information to fetch their values in Conjur and update the K8s Secrets.
+
+This enhancement can be used in the current *init solution* and will help build the foundation for later Milestones.
+
+### Design
+
+|      | Fetch/Update K8s Secret                                      | Pros                                                         | Cons                                                         | UX                                                           | Effort estimation |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ----------------- |
+| 1    | Get K8s secrets with the names specified in `K8S_SECRETS`    | - No code changes  - Backwards compatible                    | - Repetitive work <br />- Not scalable                       | List K8S Secrets one-by-one                                  | Free              |
+| 2    | Read all K8s Secrets that are marked with labels specified in `K8S_SECRETS_LABEL` env var | - Scalable grouping  - Less intervention on our part as a customer can define which types of secrets we should parse  - K8s native, as-is solution | - Not general solution so we are still K8s bound - Demands SA has "list" privileges on secrets | Add "list" to SA `verbs: ["get", "update", "list"]` <br />- List label grouping in Secrets Provider manifest<br />- Label key `conjur-secrets-label` will need to be added to each K8s Secret | 5 days            |
+
+*Decision*: Solution #2, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs. 
+
+### Customer Experience 
+
+A customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    conjur-secrets-label: prod # label name can be configurable
```

Comment:
do we need the word `label`? can't it be implicit `conjur-secrets`

Reply:
Yes! It can be anything. I wanted `-label` to make it clear we are looking for labels and not the actual name of the K8s Secret. Which do you feel is most representative? I think there might be a problem with `conjur` in the name because of the DAP offering.
@InbalZilberman what say you?

Reply:
well, we already have `conjur-map` so that ship has sailed.
i would go with `conjur-secrets`